### DISCUSSION TITLE
Logout Single Device

### DISCUSSION THREAD

**Discussion by dbubenheim** [keycloak/keycloak#28730] [Q&A]
# Logout Single Device

tldr: Is there a way to configure Keycloak to only logout a single device?

## Scenario

A user is logged in on two devices, e.g an Android smartphone and an Android TV.
Both devices were logged in using the same browser on the smartphone (Android App via custom Chrome tab, Android TV via device code).

Now, if the user logs out on the smartphone, the TV client session is logged out as well.

What we need is a way to revoke a user session only on a single client device.
I.e. logging out on the smartphone should kill THAT session on THAT device for THAT user only.
The TV client session should continue to be logged in.

### Current Behaviour

Unfortunately, neither `/logout` nor `/revoke` offer what we need as far as we know:

- `/logout` endpoint kills a user session and related client sessions
- `/revoke` endpoint kills all sessions of a specific client on all devices

| device, client  | d1, c1        | d1, c2 (offline)                                  | d2, c1    | d2, c2 (offline) |
|-----------------|---------------|---------------------------------------------------|-----------|------------------|
| `/revoke` d1 c2 | logged in.    | logged out                                        | logged in | logged out ⚠️    |
| `/logout` d1 c2 | logged out ⚠️ | logged out (session closed, but can just refresh) | logged in | logged in.       |

The warning signs highlight logouts that we would like to prevent.

### Desired Behaviour

Here is the behaviour we're looking for:

| device, client  | d1, c1        | d1, c2 (offline).                   | d2, c1    | d2, c2 (offline) |
|-----------------|---------------|-------------------------------------|-----------|------------------|
| `/revoke` d1 c2 | logged in     | logged out                          | logged in | logged in ✔️     |
| `/logout` d1 c2 | logged in ✔️  | logged out (no refresh possible) ✔️ | logged in | logged in.       |

### Considerations

We also considered the following alternatives:

* A client just deletes the token set. We discarded this because that would keep the session in the DB until it's cleaned up by the session-idle-time (which is unusually high: 300days).

Any ideas how we can achieve the desired behaviour?

**Comment by kherock** [comment]
From #27815, I've tested that you can indeed use the legacy backchannel logout method to sign out of an individual offline session. You can use this instead of `/revoke` to prevent the logout of d2,c2 in the first row.

Would appreciate input from a maintainer on whether or not an API exists for deleting an Offline session by its ID on the admin API. It only works for Regular SSO sessions, despite the Offline session sharing the same session ID. When attempting to delete an entry for an Offline session from the admin UI, Keycloak first deletes the Regular SSO if it exists, and then returns 404 on subsequent requests.

I currently use offline tokens to allow apps—primarily native mobile applications—to stay signed in indefinitely. It would be nice if at the client level, I could configure the following:

* Whether or not signing out also revokes offline tokens associated with the logged out session's ID
  * I did discover that Keycloak respects the `events.revoke_offline_access` claim in the back-channel logout token. Unfortunately, I have no control over the signing of this token, so I'd rather this appear as client-level configuration.
* Whether or not revoking an individual offline token also revokes consent

---

I have a similar use case mainly focused on end-user device management. I'd like to provide the following experiences within a platform of apps:
 
1. Sign out of all devices everywhere
1. Sign out of all clients associated with a browser window
    a. Not a must-have, but I am interested in a dynamic prompt asking if the operation should sign out of all devices. This would include smart devices or PWAs that received offline tokens separate from the SSO session.
1. Sign out/remove account from a native app
1. Sign out (or force re-authentication) on untrusted devices following a password reset

Are any of the above possible? It seems like (2) and (3) come with gotchas due to limitations on offline session management.

**Comment by AnthyG-epias** [comment]
I think this might relate to https://github.com/keycloak/keycloak/discussions/27815 

At least regarding logging out on a specific device.
